contracts/modules/TransferManager/VolumeDumpingRestrictionTM.sol

```diff
@@ -0,0 +1,366 @@
+        );
+
+        for (uint256 i = 0; i < _userAddresses.length; i++) {
+            addDumpingRestriction(_userAddresses[i], _percentsAllowed[i], _startTimes[i], _endTimes[i], _rollingPeriods[i]);
+        }
+    }
+
+    /**
+    * @notice lets the admin remove a volume dumping restriction.
+    * @param _userAddress User address to remove the volume dumping restriction
+    */
+    function removeRestriction(address _userAddress) public withPerm(ADMIN) {
+        VolumeRestriction memory toRemove = volumeRestriction[_userAddress];
+
+        // check if entry is in the past
+        // not allowed to delete past entries
+        require(toRemove.endTime > now, "Cannot remove past restrictions");
+
+        uint256 length = volumePeriodIds[_userAddress].length;
+       
+        // delete all the volume tally period ids for the user address
+        for( uint256 i = 0; i < length; i++ ){
+            uint256 periodId = volumePeriodIds[_userAddress][i];
+            delete volumeTally[_userAddress][periodId];
+        }
+
+        // reset the period ids
+        volumePeriodIds[_userAddress].length = 0;
+
+        delete volumeRestriction[_userAddress];        
+        
+        emit RemoveVolumeDumping(
+            _userAddress,
+            toRemove.percentAllowed,
+            toRemove.startTime,
+            toRemove.endTime,
+            toRemove.rollingPeriod
+        );
+    }
+
+    /**
+    * @notice lets the admin remove multiple volume dumping restriction.
+    * @param _userAddresses List of User addresses to remove the volume dumping restriction
+    */
+    function removeRestrictionMulti(address[] _userAddresses) public withPerm(ADMIN) {
+        // ensure the the array length is > 0
+        require(_userAddresses.length > 0, "Invalid array length");
+
+        for (uint256 i = 0; i < _userAddresses.length; i++) {
+            removeRestriction(_userAddresses[i]);
+        }
+    }
+    
+    /**
+    * @notice lets the admin modify volume dumping restriction for an addresses.
+    * @param _userAddress Address of the user to apply the volume dumping restriction
+    * @param _percentAllowed Percent of tokens balance allowed to transfer within a rolling period
+    * @param _startTime When the dumping restriction, 0 means now in seconds
+    * @param _endTime When the dumping restriction ends in seconds
+    * @param _rollingPeriod is the time period in seconds
+    */
+    function modifyVolumeDumpingRestriction( 
+        address _userAddress, 
+        uint256 _percentAllowed, 
+        uint256 _startTime, 
+        uint256 _endTime, 
+        uint256 _rollingPeriod
+        ) public withPerm(ADMIN) {
+
+        if (_startTime == 0) {
+            _startTime = now;
+        }
+
+        _checkVolumeDumpingParams(_userAddress, _percentAllowed, _startTime, _endTime, _rollingPeriod);
+ 
+        require(volumeRestriction[_userAddress].endTime > now, "Cannot modify past restrictions");
```

Comment:
same if the issuer wants to modify the details for a particular investor. He can't do it until the dumping period gets completed. But please do care with your logic because if Issuer changes details during the dumping period then it will affect the calculation of rolling period. 
Two ways to mitigate - 
1. Allow the ADMIN to modify only when startTime < now.
2. Change the logic of periodID to make it more dynamic.

Reply:
Okay then, I would add a restriction to allow to modify only when `startTime < now`

Reply:
Can we clarify using the below example

For example if a user has the following config 
percentAllowed = 10
rollingPeriod = 30

And then intends to change it to
percentAllowed = 20
rollingPeriod = 30

It doesn't really affect anything

But if he changes it to
percentAllowed = 10
rollingPeriod = 40

If the rolling period has started then definitely it would affect the calculation, since the rolling period is used in calculating the period id, 
Hence an approach would be to check the already spent amount in the current active auction then roll over the spent amount to the new periodId calculated.

That would seem to take care of the edge case

Or is there anything else I am missing?

Reply:
Yes, you are covering all the base but I am not a fan of deleting the previous data because if any reason Issuer wants to get the details from the contract then Issuer receives fewer data.

Reply:
Yeah another approach would be to delete the only the periodIds greater than the newPeriodID, to prevent the id collision

Reply:
Or we can prevent changing the rollingPeriod & startTime since that's what is used in calculating the periodId once the restriction period has started